stlab/concurrency/tuple_algorithm.hpp

```diff
@@ -143,6 +151,115 @@ auto void_i(T& t, std::size_t index, F&& f)
 
 /**************************************************************************************************/
 
+namespace detail {
+
+/**************************************************************************************************/
+
+template <class F, class Tuple, std::size_t... I>
+constexpr decltype(auto) apply_impl(F&& f, Tuple&& t, std::index_sequence<I...>)
+{
+    return std::forward<F>(f)(std::get<I>(std::forward<Tuple>(t))...);
+}
+
+template <class F, class Tuple, std::size_t... I>
+constexpr decltype(auto) apply_optional_indexed_impl(F&& f, Tuple&& t, std::index_sequence<I...>)
+{
+    return std::forward<F>(f)(*std::get<I>(std::forward<Tuple>(t))...);
+}
+
+/**************************************************************************************************/
+
+} // namespace detail
+
+/**************************************************************************************************/
+
+template <class Seq, class F, class Tuple>
+constexpr decltype(auto) apply_indexed(F&& f, Tuple&& t) {
+    return detail::apply_impl(std::forward<F>(f), std::forward<Tuple>(t), Seq());
+}
+
+template <class F, class Tuple>
+constexpr decltype(auto) apply_tuple(F&& f, Tuple&& t) {
+    return detail::apply_impl(std::forward<F>(f),
+                              std::forward<Tuple>(t),
+                              std::make_index_sequence<std::tuple_size<Tuple>::value>());
+}
+
+template <class Seq, class F, class Tuple>
+constexpr decltype(auto) apply_optional_indexed(F&& f, Tuple&& t) {
+    return detail::apply_optional_indexed_impl(std::forward<F>(f), std::forward<Tuple>(t), Seq());
+}
+
+/**************************************************************************************************/
+
+class placeholder { };
```

Comment:
It seems to me that from here to the next comment line, the code belongs more into the `detail` namespace, than into the `stlab` namespace. Or am I wrong? If I am wrong, then it would be good to add some lines of documentation to it.

Reply:
I made a new test executable for the feature, called `tuple_test`. I have a couple tuple-specific tests in there, as well as some that make sure the new `when_all` and `when_any` changes are tested. If you can point me to a better file for the `when_*` tests I can move them there. Let me know.

Reply:
@sean-parent Can you comment on `remove_placeholder` and `map_placeholder`, especially the nested `function` struct? I'll take whatever you've got to give and will put it in the header.

Reply:
I'll move it to detail for now. We can talk about moving out of detail, perhaps for a subsequent release.

Reply:
@fosterbrereton There are the tests for when_any and when_all in test/when_all_arguments_test and test/when_any_arguments_test.